libraries/bot-ai-qna/src/main/java/com/microsoft/bot/ai/qna/dialogs/QnAMakerDialog.java

```diff
@@ -905,27 +910,29 @@ private static void resetOptions(DialogContext dc, QnAMakerDialogOptions dialogO
                 result.add(response.getAnswers()[0]);
             }
 
-            stepContext.setValue(ValueProperty.QNA_DATA, result);
+            stepContext.getValues().put(ValueProperty.QNA_DATA, result);
             ObjectPath.setPathValue(stepContext.getActiveDialog().getState(), OPTIONS, dialogOptions);
 
             // If card is not shown, move to next step with top QnA response.
-            return stepContext.next(result);
+            return stepContext.next(result).join();
         });
     }
 
     private CompletableFuture<DialogTurnResult> callTrain(WaterfallStepContext stepContext) {
-        QnAMakerDialogOptions dialogOptions = ObjectPath.getPathValue(stepContext.getActiveDialog().getState(), OPTIONS);
-        List<QueryResult> trainResponses = (List<QueryResult>) stepContext.getValue(ValueProperty.QNA_DATA);
-        String currentQuery = (String) stepContext.getValue(ValueProperty.CURRENT_QUERY);
+        QnAMakerDialogOptions dialogOptions = ObjectPath.getPathValue(stepContext.getActiveDialog().getState(), OPTIONS,
+            QnAMakerDialogOptions.class);
+        List<QueryResult> trainResponses = (List<QueryResult>) stepContext.getValues().get(ValueProperty.QNA_DATA);
+        String currentQuery = (String) stepContext.getValues().get(ValueProperty.CURRENT_QUERY);
 
         String reply = stepContext.getContext().getActivity().getText();
 
         if (trainResponses.size() > 1) {
-            QueryResult qnaResult = trainResponses.stream().filter(kvp -> kvp.getQuestions()[0] == reply).findFirst().orElse(null);
+            QueryResult qnaResult = trainResponses
+                .stream().filter(kvp -> kvp.getQuestions()[0] == reply).findFirst().orElse(null);
             if (qnaResult != null) {
                 List<QueryResult> queryResultArr = new ArrayList<QueryResult>();
-                stepContext.setValue(ValueProperty.QNA_DATA, queryResultArr.add(qnaResult));
-                FeedbackRecord record = new FeedbackRecord(){{
+                stepContext.getValues().put(ValueProperty.QNA_DATA, queryResultArr.add(qnaResult));
```

Comment:
Is this ok? Assuming that `getValues()` returns a list or array, it will allow to put values into it, but that is not being assigned anywhere. Unless `getValues()` works by reference?

Reply:
`getValues` returns a `Map<String, Object>` that is in the stepContext. That line adds a new value to that Map, there is no need to assignee it to something

Reply:
I added the same feedback comparing the behavior against C#